### PR TITLE
o/snapstate, o/hookstate, o/ifacestate: create component hook tasks

### PR DESCRIPTION
**Pull request by andrewphelpsj**
This PR creates and enables us to run component hook tasks.

Here is the set of tasks created when installing a snap with one component:
```
1 prerequisites, waits on []
2 download-snap, waits on [1]
3 validate-snap, waits on [2]
4 mount-snap, waits on [3]
5 download-component, waits on [4]
6 validate-component, waits on [5]
7 mount-component, waits on [6]
10 copy-snap-data, waits on [7]
11 setup-profiles, waits on [10]
12 link-snap, waits on [11]
8 link-component, waits on [7 12]
13 auto-connect, waits on [8]
14 set-auto-aliases, waits on [13]
15 setup-aliases, waits on [14]
16 run-hook[install], waits on [15]
9 run-hook[install] (component), waits on [8 16]
17 run-hook[default-configure], waits on [9]
18 start-snap-services, waits on [17]
19 run-hook[configure], waits on [1 2 3 4 5 6 7 10 11 12 8 13 14 15 16 9 17 18]
20 run-hook[check-health], waits on [1 2 3 4 5 6 7 10 11 12 8 13 14 15 16 9 17 18 19]
```

Here is the set of tasks when refreshing a snap with one component:
```
1 prerequisites, waits on []
2 download-snap, waits on [1]
3 validate-snap, waits on [2]
4 mount-snap, waits on [3]
5 download-component, waits on [4]
6 validate-component, waits on [5]
7 mount-component, waits on [6]
8 run-hook[pre-refresh] (component), waits on [7]
11 run-hook[pre-refresh], waits on [8]
12 stop-snap-services, waits on [11]
13 remove-aliases, waits on [12]
14 unlink-current-snap, waits on [13]
15 copy-snap-data, waits on [14]
16 setup-profiles, waits on [15]
17 link-snap, waits on [16]
9 link-component, waits on [8 17]
18 auto-connect, waits on [9]
19 set-auto-aliases, waits on [18]
20 setup-aliases, waits on [19]
21 run-hook[post-refresh], waits on [20]
10 run-hook[post-refresh] (component), waits on [9 21]
22 start-snap-services, waits on [10]
23 cleanup, waits on [22]
24 run-hook[configure], waits on [1 2 3 4 5 6 7 8 11 12 13 14 15 16 17 9 18 19 20 21 10 22 23]
25 run-hook[check-health], waits on [1 2 3 4 5 6 7 8 11 12 13 14 15 16 17 9 18 19 20 21 10 22 23 24]
```

### CODE REVIEW

File: overlord/snapstate/component.go [hunk 1]
```diff
@@ -209,17 +239,28 @@ func doInstallComponent(st *state.State, snapst *SnapState, compSetup *Component
...
+	}
+
+	componentTS.postOpHookAndAfter = append(componentTS.postOpHookAndAfter, postOpHook)
+	addTask(postOpHook)
+
+	componentTS.ts.MarkEdge(prepare, BeginEdge)
+	componentTS.ts.MarkEdge(linkSnap, MaybeRebootEdge)
+	componentTS.ts.MarkEdge(postOpHook, PostOpHookEdge)
 
 	// TODO do we need to set restart boundaries here? (probably
```
Comment (pedronis) [inline overlord/snapstate/component.go:260]: unrelated but seems this needs to become a TODO:COMPS todo? maybe worth checking other TODOs in this file (cc @alfonsosanchezbeato)

File: overlord/snapstate/component.go [hunk 2]
```diff
@@ -156,48 +178,56 @@ func doInstallComponent(st *state.State, snapst *SnapState, compSetup *Component
...
 		}
 	}
 
-	// TODO hooks for components
+	compInstalled := snapst.IsComponentInCurrentSeq(compSi.Component)
+
+	if !snapsup.Revert && compInstalled {
+		preRefreshHook := SetupPreRefreshComponentHook(st, snapsup.InstanceName(), compSi.Component.ComponentName)
+		componentTS.beforeLink = append(componentTS.beforeLink, preRefreshHook)
+		addTask(preRefreshHook)
```
Comment (pedronis) [inline overlord/snapstate/component.go:191]: wondering whether it make sense to make this combo a variant of the addTask local helper? another approach would be to have a flag about being before that is set to false at the right point? but maybe is too implicit?

File: overlord/snapstate/component_install_test.go [hunk 3]
```diff
@@ -47,41 +47,60 @@ const (
 	compCurrentIsDiscarded
 	// Component is being installed with a snap, so skip setup-profiles
 	compOptSkipSecurity
+	// Component is being installed with a snap that is being refreshed
+	compOptDuringSnapRefresh
+	// The snap that the component is associated with is being reverted
+	compOptIsRevert
```
Comment (pedronis) [inline overlord/snapstate/component_install_test.go:53]: is this used yet?

File: overlord/snapstate/snapstate.go [hunk 4]
```diff
@@ -758,6 +724,32 @@ func doInstall(st *state.State, snapst *SnapState, snapsup SnapSetup, compsups [
...
+		componentTS, err := doInstallComponent(st, snapst, compsup, snapsup, fromChange)
+		if err != nil {
+			return nil, nil, nil, nil, fmt.Errorf("cannot install component %q: %v", compsup.CompSideInfo.Component, err)
+		}
+
+		compSetupIDs = append(compSetupIDs, componentTS.compSetupTask.ID())
+
+		tasksBeforeCurrentUnlink = append(tasksBeforeCurrentUnlink, componentTS.beforeLink...)
+		tasksAfterLinkSnap = append(tasksAfterLinkSnap, componentTS.linkToHook...)
+		tasksAfterPostOpHook = append(tasksAfterPostOpHook, componentTS.postOpHookAndAfter...)
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go:748]: this looks nice

**Review by pedronis — COMMENTED** [review]
thanks, did a pass, could you put example sequence of tasks as produced by the code for install and refresh vs component hooks in the PR description

File: tests/main/component-refreshes/task.yaml [hunk 5]
```diff
@@ -0,0 +1,40 @@
+summary: Test installing a component from the store
```
[the remark this answers is not in the document]
Comment (andrewphelpsj) [reply tests/main/component-refreshes/task.yaml:1]: It really should have been a part of the last PR. Would you like it split out?

File: snap/component.go [hunk 6]
```diff
@@ -282,9 +282,13 @@ func addAndBindImplicitComponentHook(componentInfo *ComponentInfo, snapInfo *Inf
 		}
 	}
 
-	// TODO: if hooks ever get slots, then unscoped slots will need to be
+	// TODO:COMPS: if hooks ever get slots, then unscoped slots will need to be
 	// bound here
 
+	if componentInfo.Hooks == nil {
+		componentInfo.Hooks = make(map[string]*HookInfo)
+	}
```
[the remark this answers is not in the document]
Comment (andrewphelpsj) [reply snap/component.go:290]: Yes, I added a test along with it: 9bf63b4803904ed2731428ccdfb3167a53cdb7f6.

File: overlord/snapstate/component.go (hunk 2 above)
Comment (andrewphelpsj) [reply overlord/snapstate/component.go:191]: I considered something like this, but I do think it is more readable to just be explicit with how we're collecting the tasks.

Comment (pedronis) [reply overlord/snapstate/component.go:191]: related: do we have a test that checks that ts vs the other list of tasks on componentTs in aggregate list all the same tasks?

File: overlord/snapstate/component_install_test.go (hunk 3 above)
Comment (andrewphelpsj) [reply overlord/snapstate/component_install_test.go:53]: Whoops, no.

**Comment by andrewphelpsj** [comment]
@pedronis I added the task lists to the description. While doing this, I realized that component pre-refresh hooks are supposed to go before snap pre-refresh hooks. This change is made now.

File: overlord/snapstate/component.go (hunk 2 above)
Comment (andrewphelpsj) [reply overlord/snapstate/component.go:191]: No, although that isn't really something I think we could do from the public API? We'd have to re-export `doInstallComponent`, which we could do if you like.

File: tests/main/component-refreshes/task.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (pedronis) [reply tests/main/component-refreshes/task.yaml]: maybe yes, if it's not too annoying

File: overlord/snapstate/component.go (hunk 2 above)
Comment (pedronis) [reply overlord/snapstate/component.go:191]: we do that for doInstall, so maybe is not a bad idea to have a few tests about doInstallComponent that look at the internal behavior/main behavior of it

**Comment by pedronis** [comment]
> @pedronis I added the task lists to the description. While doing this, I realized that component pre-refresh hooks are supposed to go before snap pre-refresh hooks. This change is made now.

that's why I suggested listing explicitly, so @alfonsosanchezbeato can also check they match the spec he wrote. Sadly our verify* functions are hard to follow in terms of what is the actual ordering

**Review by pedronis — COMMENTED** [review]
thanks for the changes, I'll wait after @alfonsosanchezbeato review for my re-review

File: overlord/snapstate/component.go [hunk 7]
```diff
@@ -78,33 +78,46 @@ func InstallComponentPath(st *state.State, csi *snap.ComponentSideInfo, info *sn
...
+	linkToHook         []*state.Task
+	postOpHookAndAfter []*state.Task
+	ts                 *state.TaskSet
+}
+
 // doInstallComponent might be called with the owner snap installed or not.
-func doInstallComponent(st *state.State, snapst *SnapState, compSetup *ComponentSetup,
-	snapsup *SnapSetup, fromChange string) (*state.TaskSet, error) {
+func doInstallComponent(st *state.State, snapst *SnapState, compSetup ComponentSetup,
+	snapsup SnapSetup, fromChange string) (componentInstallTaskSet, error) {
```
Comment (alfonsosanchezbeato) [inline overlord/snapstate/component.go:104]: Are there unit tests that check that the tasks are generated for components that do have the hooks?

File: tests/main/component-refreshes/task.yaml [hunk 8]
```diff
@@ -0,0 +1,40 @@
+summary: Test refreshing a component from the store
+
+details: |
+  Verifies that we can refresh a snap with components from the store.
+
+systems: [ubuntu-16.04-64, ubuntu-18.04-64, ubuntu-2*, ubuntu-core-*, fedora-*]
+
+restore: |
+  snap refresh --unhold
+  # TODO:COMPS: properly remove the snap and components
```
Comment (alfonsosanchezbeato) [inline tests/main/component-refreshes/task.yaml:10]: I think that if not removed properly it will cause trouble to the next test. But the remove PRs are about to be merged, so once that happens this PR can be rebased and the clean-up code should do the right thing simply by removing the snaps.

File: tests/main/component-refresh-hooks/task.yaml [hunk 9]
```diff
@@ -0,0 +1,48 @@
+summary: Test running component install, pre-refresh, and post-refresh hooks
+
+details: |
+  Tests a snap with two components, with two components. Some arbitrary
```
Comment (alfonsosanchezbeato) [inline tests/main/component-refresh-hooks/task.yaml:4]: Repeated string :slightly_smiling_face:

**Review by alfonsosanchezbeato — COMMENTED** [review]
Some minor comments. Re: order of the component hook tasks it looks in agreement with the spec.

File: overlord/snapstate/component.go (hunk 7 above)
Comment (andrewphelpsj) [reply overlord/snapstate/component.go:104]: Yes, since the tasks are always created regardless if the component has hooks or not, all of the tests have to consider the hook tasks.

File: overlord/snapstate/component.go [hunk 10]
```diff
@@ -78,33 +78,46 @@ func InstallComponentPath(st *state.State, csi *snap.ComponentSideInfo, info *sn
...
 	RemoveComponentPath bool `json:"remove-component-path,omitempty"`
 	SkipProfiles        bool `json:"skip-profiles,omitempty"`
 }
 
+type componentInstallTaskSet struct {
+	compSetupTask      *state.Task
+	beforeLink         []*state.Task
+	linkToHook         []*state.Task
+	postOpHookAndAfter []*state.Task
+	ts                 *state.TaskSet
```
[the remark this answers is not in the document]
Comment (andrewphelpsj) [reply overlord/snapstate/component.go:99]: I like that idea, thanks!

File: overlord/snapstate/snapstate.go [hunk 11]
```diff
@@ -79,6 +79,7 @@ const (
 	AfterMaybeRebootWaitEdge         = state.TaskSetEdge("after-maybe-reboot-wait")
 	LastBeforeLocalModificationsEdge = state.TaskSetEdge("last-before-local-modifications")
 	EndEdge                          = state.TaskSetEdge("end")
+	PostOpHookEdge                   = state.TaskSetEdge("post-op-hook-edge")
```
[the remark this answers is not in the document]
Comment (andrewphelpsj) [reply overlord/snapstate/snapstate.go:82]: This actually is leftover from a previous attempt at doing things. It is gone now.

File: overlord/snapstate/component.go (hunk 2 above)
Comment (andrewphelpsj) [reply overlord/snapstate/component.go:204]: I've reworked this with a suggestion from @alfonsosanchezbeato so that I think it would be harder for them to get out of sync, in b8a07ddfef10220e3a431c4c160a51abf4cc83c9.

File: overlord/snapstate/component.go [hunk 12]
```diff
@@ -96,7 +96,23 @@ type componentInstallTaskSet struct {
...
+	tasks = append(tasks, c.linkToHook...)
+	tasks = append(tasks, c.postOpHookAndAfter...)
+
+	if len(c.linkToHook) == 0 {
+		return nil, errors.New("internal error: link component task set does not have a link task")
+	}
+
+	ts := state.NewTaskSet(tasks...)
+	ts.MarkEdge(c.compSetupTask, BeginEdge)
+	ts.MarkEdge(c.linkToHook[0], MaybeRebootEdge)
```
Comment (alfonsosanchezbeato) [inline overlord/snapstate/component.go:113]: Thinking about it, probably we don't want to reboot when kernel-modules components are installed and maybe this edge should be removed for the moment.

**Review by alfonsosanchezbeato — COMMENTED** [review]
Just one comment

File: overlord/snapstate/component.go (hunk 12 above)
Comment (andrewphelpsj) [reply overlord/snapstate/component.go]: Edge is removed now.

**Review by alfonsosanchezbeato — APPROVED** [review]
LGTM, thanks for the changes

File: overlord/snapstate/component.go [hunk 13]
```diff
@@ -209,17 +247,24 @@ func doInstallComponent(st *state.State, snapst *SnapState, compSetup *Component
...
 
-	installSet := state.NewTaskSet(tasks...)
-	installSet.MarkEdge(prepare, BeginEdge)
-	installSet.MarkEdge(linkSnap, MaybeRebootEdge)
+	var postOpHook *state.Task
+	if !compInstalled {
+		postOpHook = SetupInstallComponentHook(st, snapsup.InstanceName(), compSi.Component.ComponentName)
+	} else {
+		postOpHook = SetupPostRefreshComponentHook(st, snapsup.InstanceName(), compSi.Component.ComponentName)
+	}
```
Comment (alfonsosanchezbeato) [inline overlord/snapstate/component.go:259]: I just noticed, `install` and `post-refresh` hooks should happen before the `prepare-kernel-modules-components` task, as they could generate kernel modules that the mentioned task needs to be aware of. Also, `discard-component` should be the last task as it is an undoable task,

File: overlord/snapstate/handlers_components.go [hunk 14]
```diff
@@ -683,6 +685,46 @@ func (m *SnapManager) doSetupKernelModules(t *state.Task, finalStatus state.Stat
 	return nil
 }
 
+func componentsToKeep(t *state.Task, compsup *ComponentSetup, snapsup *SnapSetup, snapst *SnapState) ([]*snap.ComponentSideInfo, error) {
```
Comment (pedronis) [inline overlord/snapstate/handlers_components.go:688]: this is really not pleasant, also remove-kernel-modules-setup could be used in various places, so we need to extra careful what it all means?

File: overlord/snapstate/component.go [hunk 15]
```diff
@@ -58,15 +58,17 @@ func InstallComponentPath(st *state.State, csi *snap.ComponentSideInfo, info *sn
...
-		InstanceKey: info.InstanceKey,
+		Base:                            info.Base,
+		SideInfo:                        &info.SideInfo,
+		Channel:                         info.Channel,
+		Flags:                           flags.ForSnapSetup(),
+		Type:                            info.Type(),
+		Version:                         info.Version,
+		PlugsOnly:                       len(info.Slots) == 0,
+		InstanceKey:                     info.InstanceKey,
+		PreUpdateKernelModuleComponents: kmodComps,
```
Comment (pedronis) [inline overlord/snapstate/component.go:71]: my preference is still slightly to set from a relevant task(s) if possible instead of from every single place

File: overlord/snapstate/snapmgr.go [hunk 16]
```diff
@@ -830,20 +835,7 @@ func Manager(st *state.State, runner *state.TaskRunner) (*SnapManager, error) {
...
-	}
-	removeKModsInUndo := func(t *state.Task, _ *tomb.Tomb) error {
-		return m.doRemoveKernelModulesSetup(t, state.UndoneStatus)
-	}
-	removeKModsInDo := func(t *state.Task, _ *tomb.Tomb) error {
-		return m.doRemoveKernelModulesSetup(t, state.DoneStatus)
-	}
-	runner.AddHandler("prepare-kernel-modules-components", setupKModsInDo, removeKModsInUndo)
-	runner.AddHandler("clear-kernel-modules-components", removeKModsInDo, setupKModsInUndo)
+	runner.AddHandler("prepare-kernel-modules-components", m.doSetupKernelModulesMany, m.undoSetupKernelModulesMany)
```
Comment (pedronis) [inline overlord/snapstate/snapmgr.go:838]: it's hard to tell just from the diff, those this has proper do-undo tests at the level of handler testing?

**Review by pedronis — COMMENTED** [review]
thank you for the changes, couple of comments/questions

File: overlord/snapstate/snapmgr.go (hunk 16 above)
Comment (andrewphelpsj) [reply overlord/snapstate/snapmgr.go:838]: Yes, `TestSetupKernelModules`, `TestSetupKernelModulesFails`, `TestRemoveKernelModulesSetup`, and `TestRemoveKernelModulesSetupFails` test the do/undo logic.

File: overlord/snapstate/component.go [hunk 17]
```diff
@@ -78,33 +80,58 @@ func InstallComponentPath(st *state.State, csi *snap.ComponentSideInfo, info *sn
...
+
+	return componentTS.taskSet(), nil
 }
 
 type componentInstallFlags struct {
-	RemoveComponentPath bool `json:"remove-component-path,omitempty"`
-	SkipProfiles        bool `json:"skip-profiles,omitempty"`
+	RemoveComponentPath    bool `json:"remove-component-path,omitempty"`
+	SkipProfiles           bool `json:"skip-profiles,omitempty"`
+	SkipKernelModulesSetup bool `json:"skip-kernel-modules,omitempty"`
```
Comment (alfonsosanchezbeato) [inline overlord/snapstate/component.go:94]: Maybe these two could be merged in one as they should be set in the same cases. Or maybe they are not needed at all as you can determine if the snap revision is changed or not, as if it changed we do not want to run these tasks. There is the other case of skipping when we want to run these tasks only once when installing multiple components but the snap does not change, but I think that in that case it would be cleaner to pass as argument the task to inject in the middle instead of creating it.

**Review by alfonsosanchezbeato — COMMENTED** [review]
Thanks for the changes, it is nice to see that actually moving the task after linking even cleans ups some bits. The other thing is the online discussion we are having on where to put the the values of the pre-link components.

File: overlord/snapstate/component.go (hunk 17 above)
Comment (alfonsosanchezbeato) [reply overlord/snapstate/component.go]: Thanks, but isn't `JointSnapComponentsInstall` the same as `!snapst.IsInstalled() || snapst.Current != snapSi.Revision`, that is, this is a snap install or a snap refresh? Can it be simplified further?

File: overlord/snapstate/handlers_components.go [hunk 18]
```diff
@@ -352,6 +380,29 @@ func (m *SnapManager) undoSetupComponent(t *state.Task, csi *snap.ComponentSideI
...
+	}
+
+	setupTask, err := snapSetupTask(t)
+	if err != nil {
+		return err
+	}
+
+	snapsup.PreUpdateKernelModuleComponents = snapst.Sequence.ComponentsWithTypeForRev(snapst.Current, snap.KernelModulesComponent)
+
+	// sinde we distinguish between nil and an empty slice, make sure to
```
Comment (alfonsosanchezbeato) [inline overlord/snapstate/handlers_components.go:395]: typo: sinde

File: overlord/snapstate/component.go (hunk 17 above)
Comment (andrewphelpsj) [reply overlord/snapstate/component.go]: No, not quite unfortunately, since we allow `snap refresh --revision=n snapname`, where `n` is the current revision of the snap. There is a similar case for installing a snap from file, where the snap is the same revision as the snap that is installed.

File: overlord/snapstate/handlers.go [hunk 19]
```diff
@@ -1432,6 +1432,10 @@ func (m *SnapManager) doUnlinkCurrentSnap(t *state.Task, _ *tomb.Tomb) (err erro
 		return err
 	}
 
+	if err := saveCurrentKernelModuleComponents(t, snapsup, snapst); err != nil {
```
Comment (alfonsosanchezbeato) [inline overlord/snapstate/handlers.go:1435]: Aren't the unlink-current-component tasks running before unlink-current-snap and therefore the state has mutated already?

File: spread.yaml
Comment (alfonsosanchezbeato) [inline spread.yaml:1]: I think this changes sneaked in unadvertindly.

**Review by alfonsosanchezbeato — COMMENTED** [review]
Thanks for the changes! I have one question and a couple of minor comments.

File: overlord/snapstate/component.go (hunk 17 above)
Comment (alfonsosanchezbeato) [reply overlord/snapstate/component.go]: Oh, right, I forgot about this case :slightly_frowning_face:

File: overlord/snapstate/handlers.go (hunk 19 above)
Comment (pedronis) [reply overlord/snapstate/handlers.go:1435]: how do we test for this in general, do we need variant of many install/refresh tests that involve specifically kernel components and not just components in general?

Comment (andrewphelpsj) [reply overlord/snapstate/handlers.go:1435]: Many of the `testInstallComponentsRunThrough` and `testUpdateWithComponentsRunThrough` variants include kernel-modules components, and test that we end up setting up/undoing set up of the correct kernel modules. I also added to the specific task handler tests cases for testing that we set the correct thing on the SnapSetup.

@alfonsosanchezbeato `unlink-current-component` will only run if we are replacing existing components, if we aren't doing that then we need to capture the list when we're unlinking the snap.

Comment (alfonsosanchezbeato) [reply overlord/snapstate/handlers.go:1435]: ok, I see

**Review by alfonsosanchezbeato — APPROVED** [review]

File: overlord/snapstate/handlers_components.go [hunk 20]
```diff
@@ -496,8 +551,12 @@ func (m *SnapManager) doUnlinkComponent(t *state.Task, _ *tomb.Tomb) (err error)
 	if err != nil {
 		return err
 	}
-	cref := compSetup.CompSideInfo.Component
 
+	if err := saveCurrentKernelModuleComponents(t, snapSup, snapSt); err != nil {
```
Comment (pedronis) [inline overlord/snapstate/handlers_components.go:555]: if I comment this out only linkCompSnapSuite.TestDoUnlinkComponent  itself fails, are we missing high level tests for this part of the functionality?

File: overlord/snapstate/snapmgr.go (hunk 16 above)
Comment (pedronis) [reply overlord/snapstate/snapmgr.go:840]: I see now we have a bunch of tasks related to kernels, where the implementation methods and the task name don't match, that's confusing. Also maybe we should do a quick pass to make sure we are happy with the task names?

**Review by pedronis — COMMENTED** [review]
thank you, couple of comments

File: overlord/snapstate/handlers_components.go (hunk 20 above)
Comment (andrewphelpsj) [reply overlord/snapstate/handlers_components.go:555]: Since we don’t yet support refreshing to a snap and losing some components (I have it as part of a future PR), this won’t be exercised yet. Want me to remove it from this PR and add it back later?

Comment (pedronis) [reply overlord/snapstate/handlers_components.go:555]: thx, it's fine to keep, but please leave a TODO about testing that you can remove later

File: overlord/snapstate/snapmgr.go (hunk 16 above)
Comment (alfonsosanchezbeato) [reply overlord/snapstate/snapmgr.go]: The tasks names (`prepare-kernel-snap`, `prepare-kernel-modules-components`) seem coherent, maybe we can change `doSetupKernelSnap`/`doSetupKernelModulesMany` to match.

Comment (andrewphelpsj) [reply overlord/snapstate/snapmgr.go]: I renamed them in 9523f18faeb7ca2acd4fdc636fb7eddff6099101 and 0937bb2ebb89551456fca69cd3b69b48aaad9592.

Comment (alfonsosanchezbeato) [reply overlord/snapstate/snapmgr.go]: Thanks! But somewhat they do not look they are around, both show the "This commit does not belong to any branch on this repository, and may belong to a fork outside of the repository." when I follow the link.

Comment (andrewphelpsj) [reply overlord/snapstate/snapmgr.go]: I think that is just what github shows when commits are in a fork of the destination repo/PR. You can see the last commit added to this branch is one of the commits there.

Comment (alfonsosanchezbeato) [reply overlord/snapstate/snapmgr.go]: Ok, I see the changes in the diff. I wonder if it would help to do rebases instead or merging master which at least to me it makes history more difficult read.

Comment (andrewphelpsj) [reply overlord/snapstate/snapmgr.go]: I don't totally disagree, but I think rebasing can mess with github's functionality to show changes from a last review. Additionally, we have in CONTRIBUTING.md to not force-push a branch once it has received reviews.

File: overlord/snapstate/snapmgr.go [hunk 21]
```diff
@@ -793,7 +793,7 @@ func Manager(st *state.State, runner *state.TaskRunner) (*SnapManager, error) {
 	runner.AddHandler("conditional-auto-refresh", m.doConditionalAutoRefresh, nil)
 
 	// specific set-up for the kernel snap
-	runner.AddHandler("prepare-kernel-snap", m.doSetupKernelSnap, m.undoSetupKernelSnap)
+	runner.AddHandler("prepare-kernel-snap", m.doPrepareKernelSnap, m.undoPrepareKernelSnap)
 	runner.AddHandler("discard-old-kernel-snap-setup", m.doCleanupOldKernelSnap, m.undoCleanupOldKernelSnap)
```
Comment (pedronis) [inline overlord/snapstate/snapmgr.go:797]: there is a naming mismatch also here

**Review by pedronis — APPROVED** [review]
thank you, there is still a naming inconsistency, could be taken care in a follow-up though, as it's already in master
